File: exam_latest/exam_LRX/work5.cpp
//
// Created by liryi on 24-11-24.
//
#include "opencv2/opencv.hpp"
#include <iostream>

using namespace cv;
using namespace std;

Mat img, inpaintMask; // 声明原图和掩码图像
Point prevPt(-1, -1); // 上一个鼠标点击点，用于绘制线条

// 鼠标回调函数，用于捕捉绘制区域
static void onMouse(int event, int x, int y, int flags, void*) {
    if (event == EVENT_LBUTTONUP || !(flags & EVENT_FLAG_LBUTTON))
        prevPt = Point(-1, -1);
    else if (event == EVENT_LBUTTONDOWN)
        prevPt = Point(x, y);
    else if (event == EVENT_MOUSEMOVE && (flags & EVENT_FLAG_LBUTTON)) {
        Point pt(x, y);
        if (prevPt.x < 0) prevPt = pt;
        line(inpaintMask, prevPt, pt, Scalar::all(255), 5, 8, 0); // 在掩码图像上绘制线条
        line(img, prevPt, pt, Scalar::all(255), 5, 8, 0); // 在原图上绘制线条
        prevPt = pt;
        imshow("image", img); // 更新显示图像
    }
}

int main() {
    // 硬编码图像路径
    string filename = "..\\images\\image5.png"; // 使用默认图像路径
    img = imread(filename, IMREAD_COLOR); // 读取图像
    if (img.empty()) { // 检查图像是否成功加载
        cout << "Couldn't open the image " << filename << endl;
        return -1;
    }

    namedWindow("image", WINDOW_AUTOSIZE); // 创建显示图像的窗口
    inpaintMask = Mat::zeros(img.size(), CV_8U); // 初始化掩码图像

    imshow("image", img); // 显示原始图像
    setMouseCallback("image", onMouse, NULL); // 设置鼠标回调函数

    while (true) {
        char c = (char)waitKey(); // 等待用户输入

        if (c == 27) // 按ESC键退出程序
            break;

        if (c == 'r') { // 按'r'键恢复原图
            inpaintMask = Scalar::all(0);
            img = imread(filename, IMREAD_COLOR); // 重新加载原图
            imshow("image", img);
        }

        if (c == 'i' || c == ' ') { // 按'i'或空格键执行修复
            Mat inpainted;
            inpaint(img, inpaintMask, inpainted, 3, INPAINT_TELEA); // 执行修复算法
            imshow("inpainted image", inpainted); // 显示修复后的图像
        }
    }

    return 0; // 程序结束
}

//逻辑：
//图像加载： 从硬编码路径读取目标图像并显示。
//掩码初始化： 创建与图像同尺寸的空白掩码，用于标记需要修复的区域。
//鼠标交互：
//鼠标拖动绘制修复区域，实时更新原图和掩码。
//按键操作：
//ESC: 退出程序。
//r: 恢复原图，清空掩码。
//i/空格: 调用 cv::inpaint 修复绘制区域，显示修复后的图像。
//结果分析：
//修复算法 INPAINT_TELEA 平滑填充用户标记区域，效果依赖输入掩码的准确性。
//按键功能灵活，适合修复不连续或杂乱区域。

File: exam_latest/exam_TB/work5.py
import cv2
import numpy as np

# 声明全局变量
img = None  # 原图
inpaintMask = None  # 掩码图像
prevPt = (-1, -1)  # 上一个鼠标点击点，用于绘制线条

# 鼠标回调函数，用于捕捉绘制区域
def onMouse(event, x, y, flags, param):
    global prevPt, img, inpaintMask
    # 如果左键释放或松开，不再记录鼠标位置
    if event == cv2.EVENT_LBUTTONUP or not (flags & cv2.EVENT_FLAG_LBUTTON):
        prevPt = (-1, -1)
    # 如果左键按下，记录当前点击位置
    elif event == cv2.EVENT_LBUTTONDOWN:
        prevPt = (x, y)
    # 如果鼠标在按下左键的情况下移动，绘制线条
    elif event == cv2.EVENT_MOUSEMOVE and (flags & cv2.EVENT_FLAG_LBUTTON):
        pt = (x, y)
        # 如果 prevPt 是初始值(-1, -1)，则将当前鼠标位置设为 prevPt
        if prevPt[0] < 0:
            prevPt = pt
        # 在掩码图像上绘制线条，作为修复区域
        cv2.line(inpaintMask, prevPt, pt, (255), 5, 8, 0)
        # 同时在原图上绘制线条
        cv2.line(img, prevPt, pt, (255, 255, 255), 5, 8, 0)
        prevPt = pt
        cv2.imshow("image", img)  # 更新显示图像

def main():
    global img, inpaintMask
    # 加载图像
    filename = r"image\image5.png"  # 设置图像路径
    img = cv2.imread(filename, cv2.IMREAD_COLOR)  # 读取图像
    if img is None:  # 检查图像是否加载成功
        print("Couldn't open the image " + filename)
        return

    # 创建窗口并初始化掩码
    cv2.namedWindow("image", cv2.WINDOW_AUTOSIZE)
    inpaintMask = np.zeros(img.shape[:2], np.uint8)  # 初始化掩码为黑色，确保是8位单通道图像

    cv2.imshow("image", img)  # 显示原始图像
    cv2.setMouseCallback("image", onMouse)  # 设置鼠标回调函数

    while True:
        key = cv2.waitKey(0)  # 等待用户输入

        # 按下 ESC 键退出程序
        if key == 27:  # ESC键退出
            break

        # 按下 'r' 键恢复原图并清空掩码
        if key == ord('r'):  # 按 'r' 键恢复原图
            inpaintMask = np.zeros(img.shape[:2], np.uint8)  # 清空掩码
            img = cv2.imread(filename, cv2.IMREAD_COLOR)  # 重新加载原图
            cv2.imshow("image", img)

        # 按下 'i' 或空格键执行修复操作
        if key == ord('i') or key == ord(' '):  # 按 'i' 或 空格键执行修复
            # 使用 Telea 算法进行修复
            inpainted = cv2.inpaint(img, inpaintMask, 3, cv2.INPAINT_TELEA)
            cv2.imshow("inpainted image", inpainted)  # 显示修复后的图像

    # 关闭所有 OpenCV 窗口
    cv2.destroyAllWindows()  # 关闭所有 OpenCV 窗口

# 执行程序
if __name__ == "__main__":
    main()
